### PR TITLE
Push nightly container image to container registry

### LINKED ISSUE #4704: Upload cilium-nightly images on each successful nightly
### Description of requested functionality

There is no easy way to consume nightly images today. Creating and uploading a container image daily allows testing the current master inside a near-production environment. In particular, a named public image means users can follow the normal upgrade guide without resorting to special builds.
It is likely best to have the name be distinct enough from the official release images that mistakes are difficult to make (i.e. is shouldn't be a tag under `cilium/cilium`).

### Example use cases

1- Test upgrading from a current release to nightly, in a real running cluster
2- Test a bugfix within a user cluster

### PR DESCRIPTION
**Pull request by nebril**
fixes #4704

This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4731)

### CODE REVIEW

**Comment by eloycoto** [comment]
Hi @nebril 

I don't think that this PR works at all, things to consider:

1) If you run that script on the final post section the vagrant box version will
not be in there, because of the following lines:

https://github.com/cilium/cilium/blob/74ff7df3914745d68b57cfffa01d06ff3624ce2a/test/test_suite_test.go#L244-L245

So that will not work.

I don't really liked that script too, I would push that to the docker-hub in a
different stage, I'm not sure if before/after the Nightly test, but something
like this:

```
stage('Nightly-Docker-Image') {
    steps {
        sh 'make docker-image'
        sh 'docker tag cilium/cilium:latest cilium/nightly:$(date +%Y-%m-%d)'
        sh 'docker push cilium/nightly:$(date +%Y-%m-d)'
    }
}
```

The ENV variables for the docker hub I'll not add that in the global ENV
section, because someone can mess up in the ginkgo test and doing bad thing.
I'll keep that credentials on that stage furthermore instead of doing some
docker login, I would use the following:

```
withDockerRegistry([ credentialsId: "DOCKER_HUB_CRED", url: "" ]) {
    sh 'make docker-image'
    sh 'docker tag cilium/cilium:latest cilium/nightly:$(date +%Y-%m-%d)'
    sh 'docker push cilium/nightly:$(date +%Y-%m-d)'
}
```
The doc for that:
https://jenkins.io/doc/pipeline/steps/docker-workflow/#withdockerregistry-sets-up-docker-registry-endpoint

Other thing to consider, is to push to dockerhub when the branch is master, if
not in any PR will trigger that, and for stable branch also will be pushed.

Finally, IMHO, the stage test should be something like this:

```
stage('Nightly-Docker-Image') {
    when {
        branch 'master'
    }
    steps {
        withDockerRegistry([ credentialsId: "DOCKER_HUB_CRED", url: "" ]) {
            sh 'make docker-image'
            sh 'docker tag cilium/cilium:latest cilium/nightly:$(date +%Y-%m-%d)'
            sh 'docker push cilium/nightly:$(date +%Y-%m-d)'
        }
    }
}
```

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

File: Jenkinsfile.nightly [hunk 1]
```diff
@@ -36,6 +36,7 @@ pipeline {
         GOPATH = "${WORKSPACE}"
         TESTDIR = "${WORKSPACE}/${PROJ_PATH}/test"
         SERVER_BOX = "cilium/ubuntu"
+        NIGHTLY_ID =  "${BUILD_NUMBER}"
```
Comment (raybejjani) [inline Jenkinsfile.nightly]: Is BUILD_NUMBER the the jenkins build number? I wonder if that can repeat if we delete builds etc.
Could we make this the date? (e.g. 20180705) so it's easier to refer to the images?

File: Jenkinsfile.nightly [hunk 2]
```diff
@@ -110,6 +111,17 @@ pipeline {
                 }
             }
         }
+        stage('Nightly-Docker-Image') {
+            steps {
+                withDockerRegistry([ credentialsId: "NIGHTLY_DOCKER_HUB_CRED", url: "" ]) {
+                    sh 'make docker-image'
```
Comment (raybejjani) [inline Jenkinsfile.nightly]: hmm, don't we already have the image built somewhere as part of the tests? It's more obvious to build it here, but it might be simpler to tag the existing one and push

Comment (raybejjani) [reply Jenkinsfile.nightly]: oh, I guess this was to @eloycoto since he suggested it

Comment (nebril) [reply Jenkinsfile.nightly]: @raybejjani unfortunately it's built inside vms which after tests are run are deleted via gingkgo framework.

First version of this PR was meant to copy the image from vm to host, but as @eloycoto pointed out before - the vms are already deleted after tests are run.

I personally believe that we should build the image on host and send it to vms so it's built only once, but this will need some bigger changes in the way our CI works (currently it's part of `vagrant provision`).

File: Jenkinsfile.nightly (hunk 1 above)
Comment (nebril) [reply Jenkinsfile.nightly]: AFAIK jenkins build number cannot be repeated even if we delete builds, but maybe @eloycoto can confirm?

File: Jenkinsfile.nightly (hunk 2 above)
Comment (raybejjani) [reply Jenkinsfile.nightly]: boo :( I guess it will build again then. shrug

File: Jenkinsfile.nightly (hunk 1 above)
Comment (raybejjani) [reply Jenkinsfile.nightly]: Oh, I meant the inevitable time jenkins crashes and we lose whatever database it keeps. In any case, the build numbers are rather opaque compared to some datestamp for nightlies. They (nightlies) have a natural name because they are fixed in time.

Comment (nebril) [reply Jenkinsfile.nightly]: I agree, will try to do it, it's not very straightforward without additional plugins

Comment (nebril) [reply Jenkinsfile.nightly]: @raybejjani I decided to use both, as it is possible that we will get two nightly builds on the same date and end up overwriting previous tag.

**Comment by nebril** [comment]
test-nightly

File: Jenkinsfile.nightly (hunk 1 above)
Comment (raybejjani) [reply Jenkinsfile.nightly]: > it's not very straightforward without additional plugins

I was worried you'd say that. I guess we can't shell out or use an environment variable, right?

> it is possible that we will get two nightly builds on the same date and end up overwriting previous tag

Good point. I guess the later one will be the "latest" for that day then? This seems reasonable. +1

File: Jenkinsfile.nightly (hunk 2 above)
Comment (eloycoto) [reply Jenkinsfile.nightly]: I mean, if you consider that can be build in the VM and get that, I'm happy with that, but: 

- We lost the stage part, and the Jenkins variables in there. We need to make changes in the Vagrantfiles. 
- The password will send to Vagrantbox, and we can mess-up, not now, but in the future. 
- With this one, it is easy to export only when happens on master. 

The thing is, if the nightly is taking 6 hours, 6 hours and 5 min will be ok from my point of view. Something that can been improved, and it'll help a lot is to not clean the build before do `make docker-image` that will help a lot in normal CI and will make things faster.

File: Jenkinsfile.nightly (hunk 1 above)
Comment (eloycoto) [reply Jenkinsfile.nightly]: Build Id is ok, if we change the job names we can change the start number. ;-)

File: Jenkinsfile.nightly (hunk 2 above)
Comment (raybejjani) [reply Jenkinsfile.nightly]: Yeah. It was more that we upload the same image we tested, in case we needed to correlate build timestamps. It's not that big a deal.

File: Jenkinsfile.nightly (hunk 1 above)
Comment (raybejjani) [reply Jenkinsfile.nightly]: Sure, but the nightly images are intended to be used. A random number is less useful when talking about them than the actual date (the point about multiple builds holds, however).

File: Jenkinsfile.nightly (hunk 2 above)
Comment (nebril) [reply Jenkinsfile.nightly]: Created an issue to track that: #4749

File: Jenkinsfile.nightly [hunk 3]
```diff
@@ -36,6 +36,7 @@ pipeline {
         GOPATH = "${WORKSPACE}"
         TESTDIR = "${WORKSPACE}/${PROJ_PATH}/test"
         SERVER_BOX = "cilium/ubuntu"
+        BUILD_NUMBER = "${BUILD_NUMBER}"
```
Comment (eloycoto) [inline Jenkinsfile.nightly]: Why you did this one?  I can't fully understand the reason.

File: Jenkinsfile.nightly [hunk 4]
```diff
@@ -110,6 +111,16 @@ pipeline {
                 }
             }
         }
+        stage('Nightly-Docker-Image') {
+            steps {
+                withDockerRegistry([ credentialsId: "NIGHTLY_DOCKER_HUB_CRED", url: "" ]) {
+                    sh 'make docker-image'
+                    sh 'docker tag cilium/cilium:latest cilium/nightly:latest'
+                    sh 'export ID=$(date +%Y-%m-%d)--$BUILD_NUMBER && docker tag cilium/nightly:latest cilium/nightly:$ID && docker push cilium/nightly:$ID'
```
Comment (eloycoto) [inline Jenkinsfile.nightly]: I don't really like this export here, what about something like this: 
```groovy
stage('Nightly-Docker-Image') {

    environment {
        TAG = "${new Date().format("yyyyMMdd")}-${BUILD_NUMBER}"
    }

    steps {
        withDockerRegistry([ credentialsId: "NIGHTLY_DOCKER_HUB_CRED", url: "" ]) {
            sh 'make docker-image'
            sh 'docker tag cilium/cilium:latest cilium/nightly:latest'
            sh 'docker tag cilium/nightly:latest cilium/nightly:${TAG}'
            sh 'docker push cilium/nightly:${TAG}'
            sh 'docker push cilium/nightly:latest'
        }
    }
}
```

**Review by eloycoto — CHANGES_REQUESTED** [review]

**Comment by nebril** [comment]
test-nightly

File: Jenkinsfile.nightly (hunk 3 above)
Comment (nebril) [reply Jenkinsfile.nightly]: Are all Jenkins variables available as env variables? I thought there was no way to use Jenkins variables within `sh` steps

File: Jenkinsfile.nightly (hunk 4 above)
Comment (nebril) [reply Jenkinsfile.nightly]: Makes sense

**Comment by nebril** [comment]
test-nightly

**Review by eloycoto — APPROVED** [review]

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
The build with only new stage worked fine: https://jenkins.cilium.io/job/Cilium-PR-Nightly-Tests-All/144/ , images are pushed here: https://hub.docker.com/r/cilium/nightly/tags/

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-me-please

**Review by eloycoto — APPROVED** [review]

**Review by aanm — COMMENTED** [review]
Couldn't we use something like https://docs.docker.com/docker-hub/builds/#remote-build-triggers? (Search for Remote Build triggers at the end)

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by nebril** [comment]
test-nightly

**Comment by raybejjani** [comment]
> Couldn't we use something like https://docs.docker.com/docker-hub/builds/#remote-build-triggers? (Search for Remote Build triggers at the end)

I didn't see an option to specify the tag the image will have. Can you give it the tag somewhere? @aanm 


Separately, @nebril @eloycoto, since we are building the image again outside the VMs anyway, it might be simpler to just have dockerhub do it. I had forgotten that we don't push images up normally (the pushes we have now are to the internal repository for each test run). Whether we build remotely or push the local image, we have a dependency on dockerhub and the connection to it. It might be more reliable, however, to make a small request versus uploading our 9GB image :/
What do you think?

* I realise I had asked about pushing up the exact same image that we tested with at some point. Since that isn't happening anyway, and won't happen for a while, doing even less work locally is probably better.

File: Jenkinsfile.nightly [hunk 5]
```diff
@@ -2,7 +2,7 @@
 
 pipeline {
     agent {
-        label 'fixed'
+        label 'eloy'
```
Comment (raybejjani) [inline Jenkinsfile.nightly]: Is this for testing? It might be safer to have it on a separate commit that can be reverted, and maybe a note on the PR saying don't merge with this commit in the PR.

File: Jenkinsfile.nightly [hunk 6]
```diff
@@ -110,6 +112,21 @@ pipeline {
                 }
             }
         }
+*/
+        stage('Nightly-Docker-Image') {
+            //when {
+            //    branch 'master'
+            //}
+            steps {
+                withDockerRegistry([ credentialsId: "NIGHTLY_DOCKER_HUB_CRED", url: "" ]) {
```
Comment (raybejjani) [inline Jenkinsfile.nightly]: Out of curiosity, where does NIGHTLY_DOCKER_HUB_CRED come from? I couldn't find it in https://github.com/cilium/Jenkins-library or in cilium

**Comment by eloycoto** [comment]
test-nightly

**Comment by eloycoto** [comment]
@nebril Updated the nodes with the correct docker-compose that does not affect other jobs. https://github.com/cilium/cilium/pull/4795

Tested this on nightly and LGTM and no issues with docker-compose.  Please move the comments and should be ready for merging.

**Comment by aanm** [comment]
@raybejjani it's on the image from the link I've pasted.

Where do you read about the 9GB image? The VMs are around that size, the docker image should be smaller.

File: Jenkinsfile.nightly (hunk 6 above)
Comment (nebril) [reply Jenkinsfile.nightly]: It's stored as a jenkins credential, we don't store it in repo for obvious reasons :D

**Comment by nebril** [comment]
@raybejjani builds in 2 minutes on jenkins host, and it weighs around 200MB

I can look into remote build triggers, but at this point nightly image is practically ready.

**Comment by nebril** [comment]
test-me-please

**Review by eloycoto — APPROVED** [review]

**Comment by nebril** [comment]
Hit #4383

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by nebril** [comment]
test-me-please

**Comment by raybejjani** [comment]
Oh, nevermind then. I thought all the images were the same and large.

**Comment by nebril** [comment]
test-me-please
